prov/verbs/src/verbs_eq.c

```diff
@@ -990,3 +1085,47 @@ int fi_ibv_eq_open(struct fid_fabric *fabric, struct fi_eq_attr *attr,
 	return ret;
 }
 
+int fi_ibv_eq_attach_domain(struct fi_ibv_eq *eq, struct fi_ibv_domain *domain)
+{
+	int async_fd = domain->verbs->async_fd;
+	struct epoll_event event;
+	int ret;
+
+	ret = fi_fd_nonblock(async_fd);
+	if (ret)
+		return ret;
+
+	memset(&event, 0, sizeof(event));
+	event.events = EPOLLIN;
+	event.data.ptr = domain;
+
+	if (epoll_ctl(eq->epfd, EPOLL_CTL_ADD, async_fd, &event))
+		return -errno;
+
+	fastlock_acquire(&eq->lock);
+	dlist_insert_tail(&domain->eq_entry, &eq->domain_list);
+	fastlock_release(&eq->lock);
+
+	domain->eq = eq;
+	ofi_atomic_inc32(&eq->ref);
+	return 0;
+}
+
+int fi_ibv_eq_detach_domain(struct fi_ibv_domain *domain)
+{
+	int async_fd = domain->verbs->async_fd;
+	struct fi_ibv_eq *eq = domain->eq;
+
+	domain->eq = NULL;
+	ofi_atomic_dec32(&eq->ref);
+
+	/* Always remove the entry from the list */
+	fastlock_acquire(&eq->lock);
+	dlist_remove(&domain->eq_entry);
+	fastlock_release(&eq->lock);
+
+	if (epoll_ctl(eq->epfd, EPOLL_CTL_DEL, async_fd, NULL))
+		return -errno;
```

Comment:
Move this first and skip rest of detach on failure, or we could access a freed domain.

Reply:
Thanks, I fixed it.